src/OpenTK/Platform/X11/XI2MouseKeyboard.cs

```diff
@@ -541,36 +551,41 @@ private unsafe static void ProcessRawMotion(XIMouse d, ref XIRawEvent raw)
             double v = 0;
             if (d.MotionX.number != -1)
             {
-                x = ReadRawValue(ref raw, d.MotionX.number);
+                if (ReadRawValue(ref raw, d.MotionX.number, out x))
+                {
+                    if (d.MotionX.mode == XIMode.Relative)
+                    {
+                        d.State.X += (int)Math.Round(x);
+                    }
+                    else
+                    {
+                        d.State.X = (int)Math.Round(x);
+                    }
+                }
```

Comment:
To me it isn't totally clear from the code why this is a fix for the issue (I'm not very familiar with XI2).
Do you have an explanation for why this fixes the issue?

Reply:
XI2 mouse messages can contain a lot of state information, including buttons, scroll wheels, position, etc., but not all of them are present in each message. Which ones are present is specified in the message's valuators mask, a bitfield.

Previously, ReadRawValue() was just returning 0 when a valuator was not present, which was another way the mouse position was resetting to 0. This "Skip missing valuators" patch changes the ReadRawValue() function to return true or false depending on whether the valuator is present. Only if the valuator is present does this then set the position or scrollwheel.

Reply:
Yeah ok, I was getting something like that from the code.

So basically, what was happening is that we read the `x` and `y` values without caring if they where there or not and defaulting them to `0` if they where not there.
Then later we just checked if the event was relative or absolute and just applied that `0` blindly.
And the issue then would have been messages that didn't contain a valid `x` or `y` value but where marked as `absolute` and would then set the position to `0`?
(relative would not have been a problem as adding `0` wouldn't change the value)

Am I correct in that understanding?

Reply:
Yes, that's correct.

And this happens more frequently when running on Wayland because it switches to relative mode when moving the mouse, and to absolute mode when clicking or scrolling.